<a href="https://colab.research.google.com/github/Quimexo/Diplo/blob/main/Entregable_%C3%BAnico_Aprendizaje_No_Supervisado.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Práctico entregable Aprendizaje No Supervisado

Georgina Flesia -- Laura Alonso Alemany



Utilizar la base de jugadores “female_players.csv” disponible en la página de Kaggle https://www.kaggle.com/datasets/stefanoleone992/fifa-23-complete-player-dataset . Consideren que en comparación con la base vista en clase, esta base no tiene los mismos jugadores (ahora mujeres 2023) y no tiene exactamente el mismo formato (a los nombres de las variables se les agregó una keyword para identificar a qué tipo de habilidad corresponde).

Con la nueva base, realizar el siguiente análisis. Notar que es más completo pero análogo al que realizamos con los datos FIFA2019 en el cursado de la materia.

Cada seccion debe ser comentada con celdas de texto. Elijan la visualización que dé mas información sobre los datos.



In [1]:
import numpy as np
import pandas as pd
pd.set_option('display.max_columns',100)
pd.set_option('display.max_rows',1000)
import itertools
import warnings
warnings.filterwarnings("ignore")
import io

In [2]:
from plotly.offline import init_notebook_mode, plot,iplot
import plotly.graph_objs as go
init_notebook_mode(connected=True)
import matplotlib.pyplot as plt
import plotly.tools as tls#visualization
import plotly.figure_factory as ff#visualization
import seaborn as sns

1- Análisis exploratorio inicial de la base.



In [6]:
url='https://raw.githubusercontent.com/Quimexo/Diplo/main/female_players.csv'

In [4]:
#url='https://raw.githubusercontent.com/DiploDatos/AprendizajeNOSupervisado/master/2024/Fifa2019/data2019.cs'

In [7]:
df=pd.read_csv(url)
df.head(5)

player_id                           player_url  fifa_version  fifa_update  \
0     227125       /player/227125/sam-kerr/230009            23            9   
1     227316  /player/227316/wendie-renard/230009            23            9   
2     226301    /player/226301/alex-morgan/230009            23            9   
3     227310  /player/227310/ada-hegerberg/230009            23            9   
4     227246    /player/227246/lucy-bronze/230009            23            9   

  fifa_update_date    short_name                      long_name  \
0       2023-01-13       S. Kerr              Samantha May Kerr   
1       2023-01-13     W. Renard      Wéndèleine Thérèse Renard   
2       2023-01-13     A. Morgan      Alexandra Morgan Carrasco   
3       2023-01-13  A. Hegerberg  Ada Martine Stolsmo Hegerberg   
4       2023-01-13     L. Bronze     Lucia Roberta Tough Bronze   

  player_positions  overall  potential    value_eur  wage_eur  age  \
0               ST       91         91  134500000.0    4000.0   28   
1               CB       91         91   89500000.0    3000.0   31   
2               ST       90         90          NaN       NaN   32   
3               ST       90         91  143000000.0    3000.0   26   
4               RB       89         89          NaN       NaN   30   

          dob  height_cm  weight_kg  league_id           league_name  \
0  1993-09-10        168         66     2216.0  Women's Super League   
1  1990-07-20        187         70     2218.0   Feminine Division 1   
2  1989-07-02        168         62        NaN                   NaN   
3  1995-07-10        176         68     2218.0   Feminine Division 1   
4  1991-10-28        171         65        NaN                   NaN   

   league_level  club_team_id  club_name club_position  club_jersey_number  \
0           1.0      116010.0  Chelsea W            ST                20.0   
1           1.0      116033.0     Lyon W           LCB                 3.0   
2           NaN           NaN        NaN           NaN                 NaN   
3           1.0      116033.0     Lyon W           RES                14.0   
4           NaN           NaN        NaN           NaN                 NaN   

  club_loaned_from club_joined_date  club_contract_valid_until_year  \
0              NaN       2020-01-01                          2024.0   
1              NaN       2006-07-01                          2026.0   
2              NaN              NaN                             NaN   
3              NaN       2014-07-01                          2024.0   
4              NaN              NaN                             NaN   

   nationality_id nationality_name  nation_team_id nation_position  \
0             195        Australia             NaN             NaN   
1              18           France        113003.0             LCB   
2              95    United States        113009.0              ST   
3              36           Norway        113007.0              ST   
4              14          England        113002.0              RB   

   nation_jersey_number preferred_foot  weak_foot  skill_moves  \
0                   NaN          Right          4            4   
1                   3.0          Right          3            2   
2                  13.0           Left          4            4   
3                  14.0          Right          4            4   
4                   2.0          Right          3            3   

   international_reputation    work_rate         body_type real_face  \
0                         5    High/High     Normal (170-)       Yes   
1                         5  High/Medium       Lean (185+)       Yes   
2                         5     High/Low            Unique       Yes   
3                         5  High/Medium  Normal (170-185)       Yes   
4                         5    High/High  Normal (170-185)       Yes   

   release_clause_eur                                        player_tags  \
0         275700000.0  #Aerial Threat, #Dr

In [15]:
df.shape

(181361, 110)

In [17]:
df.columns

Index(['player_id', 'player_url', 'fifa_version', 'fifa_update',
       'fifa_update_date', 'short_name', 'long_name', 'player_positions',
       'overall', 'potential',
       ...
       'cdm', 'rdm', 'rwb', 'lb', 'lcb', 'cb', 'rcb', 'rb', 'gk',
       'player_face_url'],
      dtype='object', length=110)

In [10]:
# Verificamos si hay filas duplicadas...
duplicados = df.duplicated()
df_duplicados = df[duplicados]
df_duplicados

Empty DataFrame
Columns: [player_id, player_url, fifa_version, fifa_update, fifa_update_date, short_name, long_name, player_positions, overall, potential, value_eur, wage_eur, age, dob, height_cm, weight_kg, league_id, league_name, league_level, club_team_id, club_name, club_position, club_jersey_number, club_loaned_from, club_joined_date, club_contract_valid_until_year, nationality_id, nationality_name, nation_team_id, nation_position, nation_jersey_number, preferred_foot, weak_foot, skill_moves, international_reputation, work_rate, body_type, real_face, release_clause_eur, player_tags, player_traits, pace, shooting, passing, dribbling, defending, physic, attacking_crossing, attacking_finishing, attacking_heading_accuracy, attacking_short_passing, attacking_volleys, skill_dribbling, skill_curve, skill_fk_accuracy, skill_long_passing, skill_ball_control, movement_acceleration, movement_sprint_speed, movement_agility, movement_reactions, movement_balance, power_shot_power, power_jumping, power_stamina, power_strength, power_long_shots, mentality_aggression, mentality_interceptions, mentality_positioning, mentality_vision, mentality_penalties, mentality_composure, defending_marking_awareness, defending_standing_tackle, defending_sliding_tackle, goalkeeping_diving, goalkeeping_handling, goalkeeping_kicking, goalkeeping_positioning, goalkeeping_reflexes, goalkeeping_speed, ls, st, rs, lw, lf, cf, rf, rw, lam, cam, ram, lm, lcm, cm, rcm, rm, lwb, ldm, ...]
Index: []

[0 rows x 110 columns]

In [11]:

#Panorama rápido del dataset
df.describe(include='all')

player_id                      player_url   fifa_version  \
count   181361.000000                          181361  181361.000000   
unique            NaN                          181361            NaN   
top               NaN  /player/227125/sam-kerr/230009            NaN   
freq              NaN                               1            NaN   
mean    234662.993907                             NaN      19.223538   
std      10659.745380                             NaN       2.060414   
min     226177.000000                             NaN      16.000000   
25%     227117.000000                             NaN      17.000000   
50%     227586.000000                             NaN      19.000000   
75%     241226.000000                             NaN      21.000000   
max     273050.000000                             NaN      23.000000   

          fifa_update fifa_update_date   short_name          long_name  \
count   181361.000000           181361       181361             181361   
unique            NaN              512         1690               1654   
top               NaN       2022-09-28  L. Williams  Samantha May Kerr   
freq              NaN              835          543                519   
mean        36.727703              NaN          NaN                NaN   
std         23.555409              NaN          NaN                NaN   
min          1.000000              NaN          NaN                NaN   
25%         16.000000              NaN          NaN                NaN   
50%         35.000000              NaN          NaN                NaN   
75%         54.000000              NaN          NaN                NaN   
max         99.000000              NaN          NaN                NaN   

       player_positions        overall      potential     value_eur  \
count            181361  181361.000000  181361.000000  4.831000e+03   
unique              382            NaN            NaN           NaN   
top                  GK            NaN            NaN           NaN   
freq              22735            NaN            NaN           NaN   
mean                NaN      76.269909      79.672967  1.668368e+07   
std                 NaN       5.709719       5.422202  2.425583e+07   
min                 NaN      49.000000      49.000000  3.250000e+05   
25%                 NaN      72.000000      76.000000  2.600000e+06   
50%                 NaN      76.000000      80.000000  6.500000e+06   
75%                 NaN      80.000000      83.000000  2.050000e+07   
max                 NaN      93.000000      95.000000  1.735000e+08   

           wage_eur            age         dob      height_cm      weight_kg  \
count   4831.000000  181361.000000      181361  181361.000000  181361.000000   
unique          NaN            NaN        1387            NaN            NaN   
top             NaN            NaN  1988-02-29            NaN            NaN   
freq            NaN            NaN        2651            NaN            NaN   
mean     818.505485      26.014099         NaN     168.756464      61.213580   
std      468.435485       3.907456         NaN       6.249204       5.563279   
min      500.000000      16.000000         NaN     150.000000      45.000000   
25%      500.000000      23.000000         NaN     164.000000      57.000000   
50%      700.000000      26.000000         NaN     170.000000      61.000000   
75%     1000.000000      29.000000         NaN     173.000000      65.000000   
max     4000.000000      41.000000         NaN     188.000000      84.000000   

          league_id           league_name  league_level   club_team_id  \
count   4876.000000                  4876        4785.0    4876.000000   
unique          NaN                     3           NaN            NaN   
top             NaN  Women's Super League           NaN            NaN   
freq            NaN                  2428           NaN            NaN   
mean    2215.473749                   NaN           1.0  116009.965135   


In [12]:

df.dtypes


player_id                           int64
player_url                         object
fifa_version                        int64
fifa_update                         int64
fifa_update_date                   object
short_name                         object
long_name                          object
player_positions                   object
overall                             int64
potential                           int64
value_eur                         float64
wage_eur                          float64
age                                 int64
dob                                object
height_cm                           int64
weight_kg                           int64
league_id                         float64
league_name                        object
league_level                      float64
club_team_id                      float64
club_name                          object
club_position                      object
club_jersey_number                float64
club_loaned_from                   object
club_joined_date                   object
club_contract_valid_until_year    float64
nationality_id                      int64
nationality_name                   object
nation_team_id                    float64
nation_position                    object
nation_jersey_number              float64
preferred_foot                     object
weak_foot                           int64
skill_moves                         int64
international_reputation            int64
work_rate                          object
body_type                          object
real_face                          object
release_clause_eur                float64
player_tags                        object
player_traits                      object
pace                              float64
shooting                          float64
passing                           float64
dribbling                         float64
defending                         float64
physic                            float64
attacking_crossing                  int64
attacking_finishing                 int64
attacking_heading_accuracy          int64
attacking_short_passing             int64
attacking_volleys                   int64
skill_dribbling                     int64
skill_curve                         int64
skill_fk_accuracy                   int64
skill_long_passing                  int64
skill_ball_control                  int64
movement_acceleration               int64
movement_sprint_speed               int64
movement_agility                    int64
movement_reactions                  int64
movement_balance                    int64
power_shot_power                    int64
power_jumping                       int64
power_stamina                       int64
power_strength                      int64
power_long_shots                    int64
mentality_aggression                int64
mentality_interceptions             int64
mentality_positioning               int64
mentality_vision                    int64
mentality_penalties                 int64
mentality_composure               float64
defending_marking_awareness         int64
defending_standing_tackle           int64
defending_sliding_tackle            int64
goalkeeping_diving                  int64
goalkeeping_handling                int64
goalkeeping_kicking                 int64
goalkeeping_positioning             int64
goalkeeping_reflexes                int64
goalkeeping_speed                 float64
ls                                 object
st                                 object
rs                                 object
lw                                 object
lf                                 object
cf                                 object
rf                                 object
rw                                 object
lam                                object
cam                                object
ram                                object
lm                                 object
lcm                                object
cm        

In [13]:
# Control
faltantes=df.isnull().sum()
faltantes


player_id                              0
player_url                             0
fifa_version                           0
fifa_update                            0
fifa_update_date                       0
short_name                             0
long_name                              0
player_positions                       0
overall                                0
potential                              0
value_eur                         176530
wage_eur                          176530
age                                    0
dob                                    0
height_cm                              0
weight_kg                              0
league_id                         176485
league_name                       176485
league_level                      176576
club_team_id                      176485
club_name                         176485
club_position                     176485
club_jersey_number                176485
club_loaned_from                  181280
club_joined_date                  176657
club_contract_valid_until_year    176576
nationality_id                         0
nationality_name                       0
nation_team_id                      3997
nation_position                     3997
nation_jersey_number                3997
preferred_foot                         0
weak_foot                              0
skill_moves                            0
international_reputation               0
work_rate                              0
body_type                              0
real_face                              0
release_clause_eur                176611
player_tags                       151198
player_traits                     109593
pace                               22741
shooting                           22741
passing                            22741
dribbling                          22741
defending                          22741
physic                             22741
attacking_crossing                     0
attacking_finishing                    0
attacking_heading_accuracy             0
attacking_short_passing                0
attacking_volleys                      0
skill_dribbling                        0
skill_curve                            0
skill_fk_accuracy                      0
skill_long_passing                     0
skill_ball_control                     0
movement_acceleration                  0
movement_sprint_speed                  0
movement_agility                       0
movement_reactions                     0
movement_balance                       0
power_shot_power                       0
power_jumping                          0
power_stamina                          0
power_strength                         0
power_long_shots                       0
mentality_aggression                   0
mentality_interceptions                0
mentality_positioning                  0
mentality_vision                       0
mentality_penalties                    0
mentality_composure                16003
defending_marking_awareness            0
defending_standing_tackle              0
defending_sliding_tackle               0
goalkeeping_diving                     0
goalkeeping_handling                   0
goalkeeping_kicking                    0
goalkeeping_positioning                0
goalkeeping_reflexes                   0
goalkeeping_speed                 158627
ls                                     0
st                                     0
rs                                     0
lw                                     0
lf                                     0
cf                                     0
rf                                     0
rw                                     0
lam                                    0
cam                                    0
ram                                    0
lm                                     0
lcm                                    0
cm                                     0
rcm                                    0
rm                     



mentality_composure	16003
Compostura: la capacidad del jugador para controlar su calma y frustración durante el partido valorada entre 1 y 99.


goalkeeping_speed 	158627
SprintSpeed: Speed rate of the player rated between 1-99.
SprintSpeed: índice de velocidad del jugador clasificado entre 1 y 99.


pace	22741
shooting	22741
passing	22741
dribbling	22741
defending	22741
physic	22741

ShortPassing: Player’s accuracy for short passes rated between 1-99.

LongPassing: Player’s accuracy for long passes rated between 1-99.

Dribbling: Dribbling skill of the player rated between 1-99. Dribbling is carrying the ball without losing while moving in one particular direction.





release_clause_eur	176611
player_tags	151198
player_traits	109593




In [14]:
# Contabilizamos todas las jugadoras
cont_jugadoras = df['player_id'].nunique()
print(f"El número de jugadoras es: {cont_jugadoras}")

El número de jugadoras es: 1635


2- Evaluación visual e intuitiva de a dos variables numéricas por vez. Sugerencia: hagan los scatterplots de las combinaciones de dos variables que tengan algún sentido para ustedes, porque todas las combinaciones de variables son muchas y no van a poder analizarlas como merecen.




3- Pregunta: ¿Se realizó alguna normalización o escalado de la base? ¿Por qué ?



4- Aplicación de clustering para encontrar grupos de jugadoras con habilidades equivalentes, por ejemplo, jugadoras que podrían intercambiarse en el caso de una lesión o cuando una jugadora está cansada. Para esto utilice como mínimo dos técnicas de clustering: por ejemplo k-medias, DBSCAN, mezcla de Gaussianas y/o alguna jerárquica. Justifiquen por qué eligen los diferentes hiper-parámetros que se puedan elegir según el método: número de clusters, medida de distancia, criterio de aglomeración…



5- Análisis cualitativo de los clusters encontrados. ¿Qué hay en cada cluster? ¿Son efectivamente equivalentes las jugadoras de un cluster, es decir, podrían cumplir el mismo rol en un equipo? Si se trata de clusters heterogéneos, ¿por qué razón pueden haber sido agrupadas las jugadoras del cluster? ¿Qué motiva las diferencias en tamaño?



6- Uso de alguna transformación (proyección, Embedding) para visualizar los resultados y/o usarla como preprocesado para aplicar alguna técnica de clustering.